hawc/apps/common/models.py

```diff
@@ -340,6 +340,119 @@ def get_maximum_depth(cls, assessment_id):
+                        issues.append(
+                            f"tag '{tag_name}' at index {adjusted_crumbs} has an invalid slug '{slug}'; should be something like '{default_slugify(tag_name)}' or '{recomputed_slug}'"
+                        )
+                else:
+                    # generate it
+                    slug = default_slugify(tag_name)
+                    tag["slug"] = slug
+            else:
+                adjusted_crumbs = prettify_indices(breadcrumbs + [idx])
+                issues.append(f"tag at index {adjusted_crumbs} is missing a name")
+
+            # recurse
+            if "children" in tag:
+                cls.validate_and_transform_tag_tree(tag["children"], issues, breadcrumbs + [idx])
+
+            # delete any extraneous keys that were passed in (could instead/also flag it as an issue?).
+            # why not flag an issue? Sometimes it's nice in JSON to change a key to "_oldname" or something,
+            # since JSON doesn't support comments. This way we don't break if the caller does something like that.
+            keys_to_delete = []
+            for key in tag:
+                if key not in valid_keys:
+                    # need to build up a list; we can't delete from the tag while we iterate through it
+                    keys_to_delete.append(key)
+
+            for key in keys_to_delete:
+                del tag[key]
+
+            # and now restructure the tag so that name/slug are keys inside a "data" object as Treebeard requires.
+            if "name" in tag:
+                del tag["name"]
+            if "slug" in tag:
+                del tag["slug"]
+            tag["data"] = {"name": tag_name, "slug": slug}
+
+            idx += 1
+
+    @classmethod
+    @transaction.atomic
+    def replace_tag_tree_in_assessment(cls, assessment, tag_tree: List[object]):
+        """
+        Replaces the tag tree for an assessment; this also removes reference/tag associations.
+
+        Args:
+            assessment (Assessment): assessment to operate on
+            tag_tree (List[object]): the user-supplied tags. This method will create the "assessment-<id>" top parent tag
+                                     and should NOT be included in the supplied argument.
+
+        Returns:
+            List[dict]: the new complete tag tree, including the "assessment-<id>" top parent tag and all id's
+        """
+        root_node = cls.get_assessment_root(assessment.id)
+
+        root_name = cls.get_assessment_root_name(assessment.id)
+        complete_tree = [{"data": {"name": root_name, "slug": root_name}, "children": tag_tree}]
+
+        root_node.delete()
+        cls.load_bulk(complete_tree, parent=None, keep_ids=False)
+        cls.clear_cache(assessment.id)
+        return cls.get_all_tags(assessment.id, json_encode=False)
+
+    @classmethod
+    @transaction.atomic
     def copy_tags(cls, copy_to_assessment, copy_from_assessment) -> Dict[int, int]:
```

Comment:
could you rewrite `copy_tags` to use this same `replace_tag_tree_in_assessment`?